File: backend/src/lib/verify.ts
// Compare what a TTS segment was asked to say against what Whisper heard. Whisper
// drops punctuation, lowercases, and normalizes spacing/diacritics, so the match
// is fuzzy: both sides are normalized to a bag of word tokens and scored by
// word-level edit distance.

// Lowercase, strip diacritics, drop punctuation/symbols, collapse whitespace, and
// split into word tokens. Digits are kept as-is (Whisper usually verbalizes them,
// but the threshold tolerates the occasional number mismatch).
export function normalizeWords(text: string): string[] {
  return text
    .normalize('NFKD')
    .replace(/[\u0300-\u036f]/g, '')
    .toLowerCase()
    .replace(/[^\p{L}\p{N}\s]/gu, ' ')
    .split(/\s+/)
    .filter(Boolean);
}

// Levenshtein distance between two token sequences (word-level).
function wordEditDistance(a: string[], b: string[]): number {
  if (a.length === 0) return b.length;
  if (b.length === 0) return a.length;
  let prev = Array.from({ length: b.length + 1 }, (_, i) => i);
  let curr = new Array<number>(b.length + 1);
  for (let i = 1; i <= a.length; i++) {
    curr[0] = i;
    for (let j = 1; j <= b.length; j++) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      curr[j] = Math.min(prev[j] + 1, curr[j - 1] + 1, prev[j - 1] + cost);
    }
    [prev, curr] = [curr, prev];
  }
  return prev[b.length];
}

// Word-level similarity in [0, 1]: 1 - (edit distance / longer length). Two empty
// strings count as a perfect match; one empty against a non-empty counts as 0.
export function wordSimilarity(expected: string, actual: string): number {
  const a = normalizeWords(expected);
  const b = normalizeWords(actual);
  if (a.length === 0 && b.length === 0) return 1;
  if (a.length === 0 || b.length === 0) return 0;
  const dist = wordEditDistance(a, b);
  return 1 - dist / Math.max(a.length, b.length);
}

File: backend/internal/lib/verify/verify.go
// Package verify compares what a TTS segment was asked to say against what
// Whisper heard. Whisper drops punctuation, lowercases, and normalizes
// spacing/diacritics, so the match is fuzzy: both sides are normalized to
// word tokens and scored by word-level edit distance.
package verify

import (
	"strings"
	"unicode"

	"golang.org/x/text/unicode/norm"
)

// NormalizeWords lowercases, strips combining diacritics (U+0300–U+036F, the
// same block Node strips), drops punctuation/symbols, and splits into word
// tokens. Digits are kept as-is.
func NormalizeWords(text string) []string {
	decomposed := norm.NFKD.String(text)
	var b strings.Builder
	b.Grow(len(decomposed))
	for _, r := range decomposed {
		switch {
		case r >= 0x0300 && r <= 0x036F:
			// combining mark — dropped
		case unicode.IsLetter(r) || unicode.IsNumber(r) || unicode.IsSpace(r):
			b.WriteRune(unicode.ToLower(r))
		default:
			b.WriteRune(' ')
		}
	}
	return strings.Fields(b.String())
}

func wordEditDistance(a, b []string) int {
	if len(a) == 0 {
		return len(b)
	}
	if len(b) == 0 {
		return len(a)
	}
	prev := make([]int, len(b)+1)
	curr := make([]int, len(b)+1)
	for i := range prev {
		prev[i] = i
	}
	for i := 1; i <= len(a); i++ {
		curr[0] = i
		for j := 1; j <= len(b); j++ {
			cost := 1
			if a[i-1] == b[j-1] {
				cost = 0
			}
			curr[j] = min(prev[j]+1, curr[j-1]+1, prev[j-1]+cost)
		}
		prev, curr = curr, prev
	}
	return prev[len(b)]
}

// WordSimilarity is word-level similarity in [0, 1]:
// 1 - editDistance/longerLength. Two empty strings are a perfect match; one
// empty against a non-empty counts as 0.
func WordSimilarity(expected, actual string) float64 {
	a := NormalizeWords(expected)
	b := NormalizeWords(actual)
	if len(a) == 0 && len(b) == 0 {
		return 1
	}
	if len(a) == 0 || len(b) == 0 {
		return 0
	}
	dist := wordEditDistance(a, b)
	return 1 - float64(dist)/float64(max(len(a), len(b)))
}
